Implemented Database per service pattern in Cart Service

GeorgeT88 opened the pull request:
first endpoint updated "addProductToCart"

GeorgeT88 committed:
Implemented Database per service pattern in Cart Service

first endpoint updated "addProductToCart"

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -16,236 +22,256 @@
+                    break;
+                }
+            }
+            List<String> productNames =
```
you are doing this logic even if you have already found the product in the cart entries. it will work but isn't optimal.
what I would do is:
`Optional<CartEntry> cartEntry= cartEntries.stream().filter(entry->productName.equals(entry.getProductName())).findFirst();

if(cartEntry.isPresent()){
            //add product in cart logic
        }else{
//add product not in cart logic
}`

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -16,236 +22,256 @@
-    @Autowired
     UserServiceClient userServiceClient;
 
     public CartDTO addProductToCart(String email, String productName, int quantity)
```
add Transactional annotation

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -16,236 +22,256 @@
+                cartEntry1.setProductName(productName);
+                cartEntry1.setQuantity(quantity);
+                cartEntry1.setCart(cart);
+                cartEntryRepository.save(cartEntry1);
```
if you add CascadeType.ALL on the cartEntries relation from the Cart, you need to invoke only cartRepository.save and it will persist/merge also the cart entries.

GeorgeT88 replied:
done!

GeorgeT88 replied:
![Capture6](https://user-images.githubusercontent.com/59165649/171792968-5183c1a0-fc65-494f-851f-438ced4ac63f.PNG)
new CartEntry wount persist if I dont save it.

florinv23 replied:
add it to the cartEntries list and it should work.

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -16,236 +22,256 @@
-        } else
-            return null;
-    }
+//    public CartDTO updateProductFromCart(String email, String productName, int quantity)
```
I think you will need these operations. why have you commented them?

GeorgeT88 replied:
Yes, these operations are needed but after implementing the new 'database per service' pattern most of it is red. So I have to take it one by one and resolve it. :)

GeorgeT88 committed:
update addProductToCart method

GeorgeT88 committed:
updated endpoints 'updateProductToCart' and 'removeProductfromCart'

GeorgeT88 committed:
updated cart endpoints

removed unused endpoints: getNumberOfActiveCarts and getAllCarts
updated endpoints: deleteCartByEmail and getCartByEmail

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
+                newCartEntry.setProductName(productName);
+                newCartEntry.setQuantity(quantity);
+                newCartEntry.setCart(cart);
+                cartEntryRepository.save(newCartEntry);
```
I think that this is not necessary because of the Cascade on the cartEntry

GeorgeT88 replied:
the new cartEntry wount persist in the DB if I remove "cartEntryRepository.save(newCartEntry);"

florinv23 replied:
but you haven't add it to the cartEntries collection. it should persist it after you add it to the cartEntries(without calling the cartEntryRepository).

GeorgeT88 replied:
you are right! done! 🥇

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
         }
+        else{
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(0);
```
why do you set it to 0 first?

GeorgeT88 replied:
removed "cartEntry.setQuantity(0);"

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
```
I think you should validate that the cart exists in order to prevent NPE a few lines later

GeorgeT88 replied:
done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
 
-        cart.setProductList(productList);
+        cart.setTotal(cart.getTotal() - initialProductTotal);
+        cart.setTotal(cart.getTotal()+(quantity*productDTO.getPrice()));
```
I think it could look better if you calculate cart new total in one line and set that value to the cart on the following line

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
     public CartDTO removeProductFromCart(String email, String productName, int quantity)
```
I don't think that makes sense quantity here. I think that for product decrease the PUT endpoint should be used. I think that this DELETE /cart/{productName} should mean removing all products of a type from a cart.

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
-                log.info("{} succesfully deleted from Product List",product.getProductName());
-
-            }
+        Cart cart = cartRepository.findByUserId(user.getId());
```
you could validate that the cart was found

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,251 @@
-            return CartMapper.cartEntityToDto(cart);
+            return cartEntityToDto(cart);
         } else
             return null;
```
it's not very nice to return null. better throw a not found exception

GeorgeT88 replied:
Done!

GeorgeT88 committed:
Code review comments adressed

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
-        return CartMapper.cartEntityToDto(cart);
+        cart.setTotal(total);
 
+        cartRepository.save(cart);
```
you could have used the cart returned after the save and not to an additional get cart by user id

GeorgeT88 replied:
Done! 🔝

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
         }
 
-        User user;
+        if (productDTO.getStock() == 0) {
```
I think you don't need this validation since you return some products to the stock. so you won't need the get product retrieval a few lines above

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
+        Cart cart = cartRepository.findByUserId(user.getId());
 
-        Long productsInCart = productList.stream().filter(p -> p.getProductName().equals(productName)).count();
+        cartEntryRepository.delete(cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().get());
```
it would be enough to remove the found cartEntry from the cartEntries list(without the need to call any cartEntryRepository method). you will need to set orphanRemoval=true on the OneToMany association.
the cartRepository.save will be enough.

GeorgeT88 replied:
I'm really astonished about your technical skills! 👏 
Done!

florinv23 replied:
thank you! don't be so astonished. I've played a little bit yesterday just to be sure that I'm not saying something wrong. :)

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        Cart cart;
+        if (cartRepository.findByUserId(user.getId()) != null) {
```
why do you call findByUserId twice?
why don't you do something like
`Cart cart = cartRepository.findByUserId(user.getId);
if(cart==null){
//throw exception
}`
//logic related to cart found

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
-        //TODO you might benefit from hibernate caching but you might also do the same DB query twice here(once when you want to see it's not null and 2nd time when you use it)
-        if (cartRepository.findByUser(user) != null) {
-            Cart cart = cartRepository.findByUser(user);
+        if (cartRepository.findByUserId(user.getId()) != null) {
```
why do you call findByUserId twice?
why don't you do something like
`Cart cart = cartRepository.findByUserId(user.getId);
if(cart==null){
//throw exception
}`
//logic related to cart found

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,241 @@
 
-        if (cartRepository.findByUser(user) != null) {
-            Cart cart = cartRepository.findByUser(user);
+        if (cartRepository.findByUserId(user.getId()) != null) {
```
why do you call findByUserId twice?
why don't you do something like
`Cart cart = cartRepository.findByUserId(user.getId);
if(cart==null){
//throw exception
}`
//logic related to cart found

GeorgeT88 replied:
Done!

GeorgeT88 committed:
second code review comments adressed

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,225 @@
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
+        List<CartEntry> entries = cart.getEntries().stream().filter(entry -> !productName.equals(entry.getProductName())).collect(Collectors.toList());
```
you could have done this a little bit better.
iterate the list to identify the entries matching productName(similar to the one above but with productName.equals()).
cart.getEntries().removeAll(matchingEntries).

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,225 @@
+        } else {
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
```
I don't think that this is needed(the cart entry, as loaded from the DB has the Cart associated).

GeorgeT88 replied:
Done! 👍

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,225 @@
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
```
you don't need to set this(if you debug this code you will see that the cart(loaded from the DB) hold reference to the same object as the one you set in this line, because it is the same that you've got a few lines above).

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,225 @@
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
```
you should check also here that the cart is not null

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,225 @@
-        if (productsInCart == 0) {
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        Cart cart = cartRepository.findByUserId(user.getId());
```
a small code improvement to this would be to do something like:
Cart cart = cartRepository.findByUserId(user.getId).orElseThrow(() -> throw new EntityNotFoundException("Cart not found in the Database!"));
but you need to change findByUserId return type to Optional<Cart>

GeorgeT88 replied:
Done! looks waaay more cleaner this way!
I have to implement this change for all Repositoty methods used in all projects.. don't I ? 🤣 🤣 🤣

florinv23 replied:
of course 😄 you elevate your coding standard 😉

GeorgeT88 committed:
third code review comments adressed

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,218 @@
-            cart.setUser(user);
+        List<CartEntry> matchingEntries = cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).collect(Collectors.toList());
+        cart.getEntries().removeAll(matchingEntries);
+        cart.setTotal(cart.getTotal() - cart.getEntries().stream().filter(entry -> productName.equals(entry.getProductName())).count() * productDTO.getPrice());
```
you should have done this before removing the entries from the cart. in this way 0 will be substracted from the total

GeorgeT88 replied:
Done!

florinv23 commented on src/main/java/com/boot/cart/service/CartService.java:
```diff
@@ -1,255 +1,218 @@
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not remove more than: " + productDTO.getStock() + " "
```
the message is a little bit misleading. the user does not remove(or it can increase the quantity) a product but is actually adding products to the cart(if the stock is under the passed value). I think that a safer approach would be to notify the user that he can't add that quantity and suggest the maximum stock.

GeorgeT88 replied:
Done!

GeorgeT88 committed:
added 'price' field

GeorgeT88 committed:
forth code review comments adressed

florinv23 commented:
well done
